map.go

```diff
@@ -0,0 +1,233 @@
+	default:
+		// Unable to acquire local lock without waiting. We are now a waiter.
+		// Check limits before waiting in a blocking fashion.
+		currentWaiters := atomic.AddInt32(&cm.waiters, 1)
+		defer atomic.AddInt32(&cm.waiters, -1)
+		if maxWaiters >= 0 && currentWaiters > maxWaiters {
+			return nil, dbmerr.NewMaxWaitersExceededError(int(maxWaiters), name)
+		}
+		select {
+		case cm.timeoutLock <- struct{}{}:
+			// Lock acquired. Will unlock when unlock is called.
+		case <-ctx.Done():
+			// timeout
+			return nil, ctx.Err()
+		}
+	}
+
+	if cm.mutex == nil {
+		m, err := allocator()
+		if err != nil {
+			// unlock our local lock
+			<-cm.timeoutLock
+			return nil, err
+		}
+		cm.mutex = m
+	}
+	lockCtx, err := cm.mutex.Lock(ctx)
+	if err != nil {
+		// unlock our local lock
+		<-cm.timeoutLock
+	}
+	return lockCtx, err
+}
+
+func (cm *countedMutex) unlock(ctx context.Context) error {
+	err := cm.mutex.Unlock(ctx)
+	// Unlock local lock even if the underlying implementation fails during unlock. The assumption
+	// is that underlying implementation should be able to handle a subsequent Lock call and provide
+	// mutual exclusion.
+	<-cm.timeoutLock
+	return err
+}
+
+// NewMutexMap allocates a new MutexMap. The passed db will be used for all database operations. options can be
+// used to customize behaviour.
+func NewMutexMap(db *sql.DB, options ...MutexMapOption) *MutexMap {
+	mmo := &mutexMapOptions{
+		maxLocalWaiters: -1,
+	}
+	for _, o := range options {
+		o(mmo)
+	}
+	if mmo.maxLocalWaiters < -1 {
+		mmo.maxLocalWaiters = -1
+	}
+	if mmo.allocator == nil {
+		mmo.allocator = dbMutexAllocator
+	}
+	return &MutexMap{
+		options:        mmo,
+		db:             db,
+		countedMutexes: make(map[string]*countedMutex),
+	}
+}
+
+// Lock locks named Mutex. If not already available for the given name, an underlying
+// Mutex will be allocated and kept for later use. In order to lock with a timeout pass ctx that
+// has a deadline.  The returned context can be used to detect if the lock is lost.
+func (mm *MutexMap) Lock(ctx context.Context, name string) (context.Context, error) {
+	cm, err := mm.acquireReference(name, true)
+	if err != nil {
+		return nil, err
+	}
+	lockExpirationCtx, err2 := cm.lock(
+		ctx,
+		func() (mutexOperations, error) {
+			return mm.options.allocator(ctx, mm.db, append(mm.options.mutexOptions, WithMutexName(name))...)
+		},
+		mm.options.maxLocalWaiters,
+		name,
+	)
+	if err2 != nil {
+		// remove lock if it's not used any longer because we had an error
+		mm.releaseReference(name)
+	}
+	return lockExpirationCtx, err2
+}
+
+// Unlock unlocks the named Mutex. Once no more references (including waiting lockers) are held for
+// the given name, the underlying Mutex is removed from an internal map.  So, it is likely that
+// new Mutex objects are frequently allocated and released.
+func (mm *MutexMap) Unlock(ctx context.Context, name string) error {
+	cm, err := mm.acquireReference(name, false)
```

Comment:
I think this call will increment the references by one over?  Lock() ref = 1, Unlock ref = 2, releaseReference ref = 1.  I don't think it will reach 0 to cleanup.